xds/internal/xdsclient/clientimpl.go

```diff
@@ -69,22 +145,57 @@ func (c *clientImpl) BootstrapConfig() *bootstrap.Config {
 	return c.config
 }
 
-// close closes the gRPC connection to the management server.
+// close closes the xDS client and releases all resources.
 func (c *clientImpl) close() {
 	if c.done.HasFired() {
 		return
 	}
 	c.done.Fire()
-	// TODO: Should we invoke the registered callbacks here with an error that
-	// the client is closed?
 
-	c.authorityMu.Lock()
+	c.topLevelAuthority.close()
 	for _, a := range c.authorities {
 		a.close()
 	}
-	c.idleAuthorities.Clear(true)
-	c.authorityMu.Unlock()
+
+	// Channel close cannot be invoked with the lock held, because it can race
+	// with stream failure happening at the same time. The latter will callback
+	// into the clientImpl and will attempt to grab the lock. This will result
+	// in a deadlock. So instead, we release the lock and wait for all active
+	// channel to be closed.
+	var channelsToClose []*xdsChannel
+	c.channelsMu.Lock()
+	for _, cs := range c.xdsActiveChannels {
+		channelsToClose = append(channelsToClose, cs.channel)
+	}
+	c.xdsActiveChannels = nil
+	c.channelsMu.Unlock()
+
+	var wg sync.WaitGroup
+	for _, c := range channelsToClose {
+		wg.Add(1)
+		go func(c *xdsChannel) {
+			defer wg.Done()
+			c.close()
+		}(c)
+	}
+	wg.Wait()
+
+	// Similarly, closing idle channels cannot be done with the lock held, for
+	// the same reason as described above.  So, we clear the idle cache in a
+	// goroutine and use a condition variable to wait on the condition that the
+	// idle cache has zero entries. The Wait() method on the condition variable
+	// release the lock and blocks the goroutine until signaled (which happens
+	// when an idle channel is removed from the cache and closed).
+	c.channelsMu.Lock()
+	c.closeCond = sync.NewCond(&c.channelsMu)
+	go c.xdsIdleChannels.Clear(true)
+	for c.xdsIdleChannels.Len() > 0 {
+		c.closeCond.Wait()
+	}
+	c.channelsMu.Unlock()
+
 	c.serializerClose()
+	<-c.serializer.Done()
```

Comment:
Optional: you could move this to the end on line 213 to allow this parallely to close with calling the xDS Server cleanups and authority cleanups.

Reply:
I don't quite understand the need to do that. In fact, i want to close the serializer as early as I can so that no more callbacks can be scheduled on it. Why would I wait longer to close the serializer?

Reply:
I meant waiting for it to be done. Doesn't that require some work (it processed up to when it sends the Done signal) so you could trigger the close, while it's doing that work do the rest of this function, and block on the c.serializer.Done() channel after doing all the other work in parallel. It's super minor.

Reply:
I see what you mean now. But I wouldn't want to do that for two reasons:
- It is super minor as you mentioned.
- Splitting it that way will make the code more confusing, since a reader might think the block of code in between the call to `c.serializerClose` and `<-c.serializer.Done()` is somehow related to those two lines. 
- Even if we add a clarifying comment, I think it would cause more cognitive load on the reader.
- Also, this is happening during xDS client close, which typically happens when a user's channel or server is being closed. So, a few micro/milli seconds of additional time here will not change things one way or the other.

Thanks.

Reply:
Yeah all these bullets are fair makes sense.